polyply/src/gen_seq.py

```diff
@@ -0,0 +1,189 @@
+    """
+    Given a `macro_str` corresponding to a molecular graph and a
+    `macro_type`, generate a graph given the specific requirements.
+
+    Parameters
+    ----------
+    macro_str:  str
+    macro_type: str
+         the type of the str; allowed are file, linear, branched,
+         random-linear
+    force_field: `:class:vermouth.forcefield.ForceField`
+
+    Returns
+    -------
+    `:class:networkx.Graph`
+    """
+    if macro_type == "from_file":
+        print(macro_str)
+        name, mol_name = macro_str.split(":")
+        macro = force_field.blocks[mol_name]
+    elif macro_type == "linear":
+        name, resname, n_blocks = macro_str.split(":")
+        macro = _macro_to_graph(resname, 1, int(n_blocks))
+    elif macro_type == "branched":
+        name, resname, branching_f, n_levels = macro_str.split(":")
+        macro = _macro_to_graph(resname, int(branching_f), int(n_levels))
+    elif macro_type == "random-linear":
+        name, n_blocks, residues = macro_str.split(":")
+        macro = _random_macro_to_graph(n_blocks, residues)
+    return name, macro
+
+def generate_seq_graph(sequence, macros, connects):
+    """
+    Given a sequence definition in terms of blocks in macros
+    generate a graph and add the edges provided in connects.
+
+    Parameters:
+    ----------
+    sequence: list
+      a list of blocks defining a squence
+    macros: dict[name]:networkx.graph
+      a dictionary of graphs defining a macro block
+    connects:
+      a list of connects
+
+    Returns:
+    -------
+    `:class:networkx.graph`
+    """
+    seq_graph = nx.Graph()
+    for idx, macro_name in enumerate(sequence):
+        sub_graph = macros[macro_name]
+        nx.set_node_attributes(
+            sub_graph, {node: idx for node in sub_graph.nodes}, "seqid")
+        seq_graph = nx.disjoint_union(seq_graph, sub_graph)
+
+    if connects:
+        for connect in connects:
+            idx, jdx, edges = connect.split(":")
+            _add_edges(seq_graph, edges, int(idx), int(jdx))
+
+    return seq_graph
+
+
+def gen_seq(args):
+
+    macros = {}
+
+    for macro_type in ["from_file", "linear", "branched", "random_linear"]:
+        macro_strings = getattr(args, macro_type)
+
+        if macro_strings:
+           if macro_type == "from_file" and macro_strings:
+               force_field = load_library("seq", args.lib, args.inpath)
+           else:
+               force_field = None
+
+           for macro_string in macro_strings:
+               name, macro_graph = interpret_macro_string(macro_string,
+                                                          macro_type,
+                                                          force_field)
+               macros[name] = macro_graph
+
+    seq_graph = generate_seq_graph(args.seq, macros, args.connects)
+
+    g_json = json_graph.node_link_data(seq_graph)
+    with open(args.outpath, "w") as file_handle:
+        json.dump(g_json, file_handle, indent=2)
```

Comment:
So this doesn't write a topology?

Reply:
No it writes a json file with sequence? Why are you asking?

Reply:
It was unclear to me what the intended output of the code is. If that's defined as json, that's fine for me.